[ci skip] Improve accessibility

SleeplessByte opened the pull request:
### Motivation / Background

As requested by @AmandaPerino, this PR aims to improve the accessibility of the guides. It continues on the work in #51341 #51372 #51385 and #51444. It also addresses whatever wasn't addressed from https://github.com/rails/rails/pull/51341#issuecomment-2007717282.

### Detail

Each commit in the PR has a descriptive title and often a message why the change was made, which I recommend we keep when merging instead of squashing it. Screenshots are listed below. A few of the commits that introduce fixes or features are listed here:

* [Add label and accessible content](https://github.com/rails/rails/commit/4f2b25ba54d82bc897fb33f681214f5bf05cb0bd) 
  
  Having a select element without a label outside a form isn't directly problematic, but the text Version is the label of the select and should act as such.

  Because the select element is set up to browse to a different page which loses the current focus position, users must be instructed about this context change. It would be much better to display this content to everyone, but having this for screenreaders only is better than not having it at all.

* [Fix accessible text for voice commands](https://github.com/rails/rails/commit/1863e5590cecac6ecfd0c4c031f1c4c967d0d768) 
  
  Using the title attribute means some screenreaders will not be able to show the expanded accessible text at all, but using aria-label loses the hover text currently available.

  If the title attribute is still used, the accessible label should start with the same content as the visible label. This is necessary for most voice operated tooling which only match starting text.

  Having "Return" as the start of the accessible label and "Guides" as the start of the visible label means in most cases users that rely on voice commands will not be able to browse here.

* [Add landmarks to all guide pages](https://github.com/rails/rails/commit/9669f4be012f34488d52e16d0b6e40a11399e414) 
  
  Landmarks such as header, section, footer, article, and the main content landmark help navigating when using assistive tooling.

  Since each guide is a self-contained article, a header section with the guide introduction and navigation, a footer section which is not part of the content but a call to action, and the rest can be clearly separated, landmarks here are appropriate.

* [Make back-to-top work without JavaScript](https://github.com/rails/rails/commit/b33278c01384a7612d0adf2270456718181b5acb) 

  By doing feature detection on animation timeline CSS, the graceful degradation can take care of the visibility of the back-to-top button. In case no JavaScript is available, it should always be shown.

* [Enable back-to-top without JavaScript](https://github.com/rails/rails/commit/430f27c959f0df30b6baeea3d688d8fa5448f842) 

  And add an accessible label to the button. It would be better to just show the text back to top underneath the arrow, but this change is again better than not having it at all.

* [Consistent scrolling behaviour respecting pref](https://github.com/rails/rails/commit/f2ae0318736a286a13d896cad8527bd2d1443c50) 

  Use the preference to detect if a user wants reduced motion and if not enable smooth scrolling. This takes care of the back-to-top smooth scrolling, and also adds it to anchor links. If this is unwanted, the change can be reverted and a matchmedia check should be done in JavaScript instead before smooth scrolling

* [Add skip links to quickly navigate past navs](https://github.com/rails/rails/commit/653b292ac26b2df59ebeb44fc9415ebc8130243b) 
  
  Skip links are a common practice to allow keyboard and other users to skip past large blocks of content that are present on every page, which in this case are the secondary navigation at the top,
the version selector, and the chapter navigation.

  The skip links only show when focused.

* [Wrap tables to make them responsive](https://github.com/rails/rails/commit/e2cfe8fdfccfcb47aec2f358b0badab79c817981)

* [Make back-to-top link to skip link](https://github.com/rails/rails/commit/ad966a42de9e166ab33026b28fc8dda64f5fbdd1) 

  This allow screen readers to announce something, because linking to an empty fragment or the body element doesn't announce anything useful.

* [Add branded selectable color](https://github.com/rails/rails/commit/2557614d872629462e354b2bd1536736aa2077be) 

  Fixes selecting text in dark mode and brands it for both modes.

* [Make more info and guides index accessible](https://github.com/rails/rails/commit/f9eaeb18d8c5712ae0f2edf351ff9829c996bf78) 

  * Adds aria-controls, which binds the control to the content.
  * Adds aria-expanded, which signals the current expansion state.
  * Adds focus handling, which focuses the first link or the button
    depending on the expansion state.
  * Adds keyboard closing for the more info button.
  * Adds outside click tracking to close the expanded content.
  * Fixes the more info button to not be focusable.

* [Fix dark mode for guide selector](https://github.com/rails/rails/commit/9c4cfd00492a2bd55db9ed5b5a9a616a962c153a)

* [Add accessible label for guide navigator](https://github.com/rails/rails/commit/93f9819c62890301f398e752b6083c9d2c9f6c40)

### Additional information

Please note that this PR doesn't necessarily make the guides completely accessible and we should continue to improve on this, but this gets us (a big) part of the way there, together with the changes that were already made prior.

| Before | After |
| ------ | ----- |
| You could not focus the selector pressing "Version" | ![a11y-focusable-label-version](https://github.com/rails/rails/assets/1964376/58f5d1a7-a6ef-47e2-86ac-6e1e0288a18c) |
| The element (version selector) was broken if JavaScript was disabled | ![a11y-no-js-desktop](https://github.com/rails/rails/assets/1964376/0a93871b-c55a-4f9b-a524-86f6024f39b7) |
| The element (version selector) was broken if JavaScript was disabled | ![a11y-no-js-mobile](https://github.com/rails/rails/assets/1964376/5075e1ac-c888-496e-8fa4-4f9066d6aa43) |
| ![a11y-before-table-wrap](https://github.com/rails/rails/assets/1964376/29e41733-60ce-4933-9d4f-36a9fad972e8) | ![a11y-after-table-wrap](https://github.com/rails/rails/assets/1964376/303818b1-d3af-45e3-9ea7-08121c91f04c) |
| You could not skip past the secondary top navigation | ![a11y-main-skip-link](https://github.com/rails/rails/assets/1964376/8d2c1af0-7575-4089-8e89-e7cfb42c8975) |
| You could not skip past the chapter navigation | ![a11y-chapter-skip-link](https://github.com/rails/rails/assets/1964376/87cccb26-f7ae-4ae9-a382-145d6cbd963e) |
| ![a11y-selectable-text-before](https://github.com/rails/rails/assets/1964376/e9e9b56f-08de-45d8-955f-c7884b33a387) | ![a11y-selectable-text](https://github.com/rails/rails/assets/1964376/c938bbde-54ee-480b-b295-6a91bdd8c7e7) |
| There was no label indicating what the select does | ![a11y-guide-selector](https://github.com/rails/rails/assets/1964376/cfaf0501-516b-4ad3-bc30-d77701f01164) |

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

SleeplessByte committed:
Add label and accessible content

Having a select element without a label outside a form isn't
directly problematic, but the text Version is the label of the
select and should act as such.

Because the select element is set up to browse to a different page
which loses the current focus position, users must be instructed
about this context change. It would be much better to display this
content to everyone, but having this for screenreaders only is
better than not having it at all.

SleeplessByte committed:
Fix accessible text for voice commands

Using the title attribute means some screenreaders will not be able
to show the expanded accessible text at all, but using aria-label
loses the hover text currently available.

If the title attribute is still used, the accessible label should
start with the same content as the visible label. This is necessary
for most voice operated tooling which only match starting text.

Having "Return" as the start of the accessible label and "Guides"
as the start of the visible label means in most cases users that
rely on voice commands will not be able to browse here.

SleeplessByte committed:
Add utilities for screen readers

This enables hiding content visually whilst still enabling it for
assisistive technologies.

SleeplessByte committed:
Feature detection for JavaScript functionalities

SleeplessByte committed:
Add landmarks to all guide pages

Landmarks such as header, section, footer, article, and the main
content landmark help navigating when using assistive tooling.

Since each guide is a self-contained article, a header section
with the guide introduction and navigation, a footer section which
is not part of the content but a call to action, and the rest can
be clearly separated, landmarks here are appropiate.

SleeplessByte committed:
Make back-to-top work without JavaScript

By doing feature detection on animation timeline CSS, the graceful
degradation can take care of the visiblity of the back-to-top
button. In case no JavaScript is available, it should always be
shown.

SleeplessByte committed:
Enable back-to-top without JavaScript

And add an accessible label to the button. It would be better to
just show the text back to top underneath the arrow, but this
change is again better than not having it at all.

SleeplessByte committed:
Consistent scrolling behaviour respecting pref

Use the preference to detect if a user wants reduced motion and if
not enable smooth scrolling. This takes care of the back-to-top
smooth scrolling, and also adds it to anchor links. If this is
unwanted, the change can be reverted and a matchmedia check should
be done in JavaScript instead before smooth scrolling

SleeplessByte committed:
Add skip links to quickly navigate past navs

Skip links are a common practice to allow keyboard and other users
to skip past large blocks of content that are present on every
page, which in this case are the secondary navigation at the top,
the version selector, and the chapter navigation.

The skip links only show when focused.

SleeplessByte committed:
Wrap tables to make them responsive

SleeplessByte committed:
Make back-to-top link to skip link

This allow screen readers to announce something, because linking
to an empty fragment or the body element doesn't announce anything
usefull.

SleeplessByte committed:
Fix false positive for broken (skip) link anchor

SleeplessByte committed:
Add branded selectable color

Fixes selecting text in dark mode and brands it for both modes.

SleeplessByte committed:
Make more info and guides index accessible

* Adds aria-controls, which binds the control to the content.

* Adds aria-expanded, which signals the current expansion state.

* Adds focus handling, which focuses the first link or the button
  depending on the expansion state.

* Adds keyboard closing for the more info button.

* Adds outside click tracking to close the expanded content.

* Fixes the more info button to not be focusable.

SleeplessByte committed:
Fix CSS for more info button

SleeplessByte committed:
Fix dark mode for guide selector

SleeplessByte committed:
Add accessible label for guide navigator

p8 commented on guides/assets/javascripts/guides.js:
```diff
@@ -24,46 +24,74 @@
+    //
+    // Additionally, keyboard users should be able to close these guides by
+    // pressing escape, which is the standard key to collapse expanded elements.
+    var guidesMenuButton = document.getElementById("guidesMenu");
+    var guides = document.getElementById(guidesMenuButton.getAttribute('aria-controls'));
```
While we are renaming things here, I think we can probably also rename the id of the button and rename the `guides` to `guidesMenu`?
```suggestion
    var guidesMenuButton = document.getElementById("guidesMenuButton");
    var guidesMenu = document.getElementById(guidesMenuButton.getAttribute('aria-controls'));
```

SleeplessByte replied:
Sure!

I tried not to touch unrelated things except for those that really didn't make sense in context.

Any other renames we should apply right away?

p8 replied:
I think these are the most obvious ones that are touched by this PR.
Any others I'm missing?

SleeplessByte replied:
I'd opt to rename container to main as it's... not a container. s-hidden also doesn't really mean anything and there is already a different class used to hide things conditionally.

Should I just make changes to improve the maintainability now that we touch these things?

jathayde commented:
This looks great @SleeplessByte. I agree on the naming `container` to something else. Open to any improvements there. I was probably overly sensitive in not rewriting it wholesale, so a lot of these things persisted when I should have pulled them.

SleeplessByte commented:
On the other hand it does add a bit of noise to the PR. I understand the rethoric in the contributing guide hehe.

I'll wait for @carlosantoniodasilva review before applying to keep it as noise free.

brunoprietog commented on guides/source/layout.html.erb:
```diff
@@ -62,7 +71,7 @@
           <li><a class="nav-item" id="home_nav" href="https://rubyonrails.org/">Home</a></li>
           <li class="guides-index guides-index-large">
-            <a href="index.html" id="guidesMenu" class="guides-index-item nav-item">Guides Index</a>
+            <a href="index.html" id="guidesMenu" aria-controls="guides" aria-expanded="false" class="guides-index-item nav-item">Guides Index</a>
```
What do you think about giving a button role to this link? It's not really navigating to any page, and that's the only purpose of a link actually. (Unless JavaScript is disabled, in which case it shouldn't have those other ARIA attributes either if that's the case).

```suggestion
            <a href="index.html" id="guidesMenu" role="button" aria-controls="guides" aria-expanded="false" class="guides-index-item nav-item">Guides Index</a>
```

SleeplessByte replied:
I like it. 

Let's apply it in JavaScript so it's only a button if... It's a button 😁

brunoprietog commented on guides/source/index.html.erb:
```diff
@@ -7,6 +7,10 @@
+    Skip to article body
+  </a>
+
   <dl>
```
This section is being read as if it were a list of items, and it's not. Let's ignore it.

```suggestion
  <dl role="presentation">
```

brunoprietog reviewed (commented):
Great work! Thank you so much for working on this. I left a few comments, several have a bias from me as a screen reader user.

SleeplessByte commented on guides/source/layout.html.erb:
```diff
@@ -94,48 +106,54 @@
-        </p>
+          <hr>
+
+          <footer aria-labelledby="heading-feedback">
```
In html 5, footer may also be an element inside an article or section. I experience the same behaviour as you (because I also use NVDA when I use a screen reader) but I believe jaws does denote it differently.

brunoprietog replied:
Oh, I don't know. I'm not too convinced about adding so many additional regions beyond the navigation, main region and footer. It adds a lot of noise that is already covered by the heading levels and becomes a bit repetitive information.

By the way, I personally don't tend to use many links to skip to specific sections and hope this can be done natively. And really, there's no way to skip to the body of the article with a command to go to a region. You can easily use 2, but that makes me think we're not being very consistent with using regions, or if we're even using them properly.

What do you think?

SleeplessByte replied:
My personal thoughts here were:

- I am browsing the guides extensively
- I am in the middle of the article and find something I want to give feedback on
- I know there is a _region_ (labelled section) that explains how to give feedback
- I want to go there straight away

With NVDA I can get there right away when browsing the article pressing <kbd>D</kbd>, which should navigate to the next landmark or region. In this case the footer is nested inside article, and thus not _complementary_, but it is labelled by the heading, and thus it becomes a section with role="region", iirc. Using the TOC navigation here would be sub-optimal as it doesn't allow be to go to "last heading of the current level", by pressing <kbd>2</kbd>. I would need to press it repeatedly, right? Since there are so little regions, that's not the case.

Does this make sense?

SleeplessByte replied:
We have a region for the article: https://github.com/rails/rails/pull/51499/files/93f9819c62890301f398e752b6083c9d2c9f6c40#diff-dc279d925a05373ac5f34307441799f98ac4accbe0b0a8f12f99f703585afa7dR113 Is that not working for you? I haven't tested mine but I can.

It _should_ be:

```html
<main>
  <article>

    <header id="feature">
      article introduction

      chapter index / toc
    </header>   

    article contents with lots of headings    

    <footer>
     <h2>Feedback</h2>

     feedback explanation
    </footer>
  </article>
</main>
```

SleeplessByte replied:
(sidenote: I am fine with explicit role-ing things if SRs don't recognise it, so let's do that if we must)

brunoprietog replied:
Right! So I think assigning the explicit roles would be enough to solve this.

In any case, an article is not exposed as a region, it's exposed as another type of element. In fact, at least in NVDA, there's no default shortcut to jump to it, I configured the A to do so but it's not the common case.

SleeplessByte replied:
Cool! Yeah, I _love_ different support for different SRs and even different versions of the same SR 🗡️ 😁. 

I'll explicitly role both the article header and article footer (so your NVDA can skip back/forth to them)

brunoprietog replied:
Yes, sadly. Just like the history of browser feature compatibility 😰

SleeplessByte commented on guides/rails_guides/markdown.rb:
```diff
@@ -148,8 +151,12 @@ def generate_index
-          <nav id="subCol">
-            <h3 class="chapter">
+          <nav id="subCol" aria-label="Chapter" class="guide-index">
+            <a id="chapter-nav-skip-link" href="#article-body" class="skip-link">
```
What does this accomplish?

I thought same-page anchors were fixed in turbo 😁

brunoprietog replied:
At least with NVDA it doesn't focus the region correctly 😢

SleeplessByte replied:
Oh weird. It does on mine, but that's good enough reason. I'll make the change!

SleeplessByte commented on guides/source/layout.html.erb:
```diff
@@ -94,48 +106,54 @@
+          <hr>
+
+          <footer aria-labelledby="heading-feedback">
+            <h3 id="heading-feedback">Feedback</h3>
```
I think it should!

SleeplessByte commented on guides/source/index.html.erb:
```diff
@@ -7,6 +7,10 @@
 
 <% content_for :index_section do %>
 <nav id="subCol" aria-label="Chapter" class="guide-index">
```
Yeah I agree. It believe the choice was made to have it as nav because on articles it does contain links but on the index it doesn't.

Let's change it here.

SleeplessByte commented on guides/source/index.html.erb:
```diff
@@ -7,6 +7,10 @@
 <% content_for :index_section do %>
 <nav id="subCol" aria-label="Chapter" class="guide-index">
+  <a id="chapter-nav-skip-link" href="#article-body" class="skip-link">
+    Skip to article body
```
Same as above

SleeplessByte commented on guides/source/layout.html.erb:
```diff
@@ -28,14 +28,22 @@
+    Skip to main content
+  </a>
+
   <nav id="topNav" aria-label="Secondary">
```
It must have a label because it's not the primary navigation (otherwise it fails a wcag criterium), as there is already a primary navigation element.

I hate the label "secondary" personally, do you have a different suggestion?

brunoprietog replied:
I understand that it's not the primary navigation, but when this element is shown, the element containing the primary navigation is hidden, so in theory this is the one that now becomes the primary navigation element. What do you think?

SleeplessByte replied:
I think you are right. Okay my proposal will be to drop the label and instead add an ERB comment for the next developer that the navs are mutually exclusive and thus not actually secondary.

jathayde replied:
Is the name something used for voice navigation or is it more to provide context? Perhaps "global" and "chapter contents" or some other delineation that alludes to the content?

SleeplessByte replied:
In our screen readers it will do something along the lines of: "Secondary navigation". It must provide context. However, the "primary" (top) navigation is allowed to not have a contextual label. That's basically what we were discussing here.

SleeplessByte replied:
@brunoprietog I actually think this can be made just a `div`. It's not a `nav` in the sense that it only holds a single button that expands a set of links. Do you agree?

SleeplessByte commented:
Thank you @brunoprietog! Very very helpful.

I left a few questions on the review and will apply all other changes.

brunoprietog commented on guides/source/layout.html.erb:
```diff
@@ -44,14 +52,15 @@
-            <select class="guides-version">
+          <a href="index.html" title="Guides home for <%= @edge.present? ? 'Edge' : @version %> Guides">Guides</a>
+          <span id="version_switcher" class="js-only">
+            <label for="version_switcher-select">Version: <span class="sr-only">pick from the list to go to that Rails version's guides</span></label>
```
Well, so I think it's fine to leave it as it's. I didn't really have a very strong opinion on that either.

SleeplessByte committed:
Only use nav on sub-column if it's actually a nav

On the index page the sub-column is not a navigational element and
does not contain content that needs to be skipped over. Therefore
it should not be a <nav> and not contain a skiplink. Furthermore,
the definition list doesn't contain any definitions, thus its role
should be removed.

SleeplessByte committed:
Ignore turbo for in-page skiplinks

This ensures the JavaScript doesn't overridde any screenreader
behaviour and the resulting fragment is correctly focused as
expected.

SleeplessByte committed:
Use keydown for escape detection

When an action should be perceived as fast, or no pause is required
before the interaction, keydown should be used and not keyup, which
can add (the feeling of a) significant delay.

SleeplessByte commented on guides/assets/javascripts/guides.js:
```diff
@@ -24,46 +24,74 @@
-    backToTop.addEventListener("click", function(e) {
-      e.preventDefault();
-      window.scrollTo({ top: 0, behavior: "smooth" });
+    document.addEventListener("keyup", function(e) {
```
In dabdb0aeef

SleeplessByte committed:
Fix heading levels

It is important that the heading structure (all hx elements) are
properly closed, and make sense. They can only ever increment by
1 since the previous heading (and decrement by many).

SleeplessByte commented on guides/rails_guides/markdown.rb:
```diff
@@ -148,8 +151,12 @@ def generate_index
+              Skip to article body
+            </a>
+
+            <h2 class="chapter">
```
In 7437e861a1

SleeplessByte committed:
Presentational picture needs aria-hidden="true"

The picture here adds a visual gimmic, but doesn't provide
information to the user and thus should be removed from the
accessibility tree.

SleeplessByte commented on guides/rails_guides/markdown.rb:
```diff
@@ -148,8 +151,12 @@ def generate_index
+            </a>
+
+            <h2 class="chapter">
               <picture>
```
In 25057200c4

SleeplessByte committed:
JS-enabled link-as-button should have correct role

When JavaScript is enabled, this link turns into a button that
expands an element. The change ensures that screenreaders only
announce it as such if JavaScript is enabled and applied.

SleeplessByte committed:
Accessible regions, focusable anchors, cleanup

Applied requested changes from review. See PR for discussion about
explicitly adding a role="region"

https://github.com/rails/rails/pull/51499

p8 commented on guides/assets/stylesrc/_dark.scss:
```diff
@@ -101,9 +105,9 @@
     }
 
-    nav#topNav {    background-color: $gray-900;
+    #mobile-navigation-bar {    background-color: $gray-900;
```
```suggestion
    #mobile-navigation-bar {
      background-color: $gray-900;
```

SleeplessByte replied:
(something something we should use an auto-formatter imho hehe)

p8 replied:
Yes :)

SleeplessByte commented on guides/source/layout.html.erb:
```diff
@@ -90,59 +102,70 @@
-  </section>
+      <div class="wrapper">
+        <div id="column-main">
+          <section id="article-body">
```
For the guides index it would have the entire index in there, which kinda makes sense. Would we label the region? Or just as a way to jump there? It is harder if we need to label it, but still possible.

brunoprietog replied:
I mean, add a region to the article body wrapper so that you can easily jump to it using quick navigation commands between regions. But only in the article body, not at the general level, because in the guide index there should be no such region.

We should probably create an extra wrapper in `markdown.rb`.

SleeplessByte replied:
The change is

```ruby
doc.wrap("<section role='region'>")
```

on line 134, and we need some changes to the anchors checking then. However, I can not make this change because after rebasing I can no longer run `rake guides:generate` probably because it expects the latest `rails` version which I cannot install on the box I am running this PR on. I therefore would like to defer this change for now.

SleeplessByte committed:
Apply requested changes from review and re-apply after rebase

SleeplessByte commented:
@AmandaPerino ready for final review

carlosantoniodasilva commented:
@SleeplessByte what have you used to auto-format the JS/SCSS files? The extra formatting is making it a bit more difficult to separate what's actual change vs not, so I'm wondering if we should push a formatting change first, and merging onto yours to make the diff less noisy.

SleeplessByte commented:
It was me accidentally not turning off format on save.

It's using the latest prettier. You can run once using `npx prettier --write .` inside the guides directory with default settings.

Apologies for that!

jathayde commented:
Is that a code standard we ascribe to across the Rails repo? I'm fine if it is, but we need to run that on everything, and then make it part of a check in process. The underscores for IDs, hyphens for classes is something I've done for 20 years. Not saying it's right, but if we are going to implement a standard, let's figure out what that is and make part of the checkboxes on PRs, etc.

carlosantoniodasilva commented:
@jathayde Rails does not have any sort of standard for those, as it's just JS/CSS used for the guides themselves. (it does have standards for the actual framework Ruby code, however.)

I assumed the hyphen changes were related to the accessibility, but I haven't gotten that far as looking into the individual commits to verify my assumption. Either way, I'd rather not make a commitment of running any automated formatting tools and come up with too many standards for the assets part of the guides, it's supposed to be a lightweight process, focused on the content. (that said, I don't mind running a formatting tool like prettier every once in a while, as it's what most people use as default for these anyway these days.)

SleeplessByte commented:
I'll pull in latest changes from #51504

zzak commented:
Thanks for your work on this!

Just putting it out there but there is already a linter and w3c validator (not currently run in CI, I believe) on the guides:
https://github.com/rails/rails/blob/965b8c372ac0e19021f92cf798095af6046ab140/guides/Rakefile#L26-L49

[[example](https://buildkite.com/rails/rails/builds/107333#018f88f6-630f-4f87-b183-14dfdb5d06e2)]

So there if there was a lightweight way to validate the a11y too, I don't see why we wouldn't take advantage of that and prevent regressions from the work done here.

zzak commented:
I don't want to make more work for you or reviewers here, apologies for the digression.
Let's chat about it after this gets merged.

brunoprietog commented:
I still want to second @SleeplessByte, automated tests might help with some minor use cases but they rarely detect serious accessibility issues.

SleeplessByte commented:
I have not forgotten about this. Will attempt to finish this weekend now that the other PRs are merged.

SleeplessByte commented:
I will do a final test with my screen reader tomorrow (Monday) and then it should be good for final review.

SleeplessByte committed:
Merge branch 'main' into accessibility

SleeplessByte committed:
Adapt the highlight in the navigation side column

for the new page structure. Additionally, replace the active class
with aria-current="true", which is meant to keep track of what is
currently active in a set. Finally, update the implementation of
the intersection observer to:

* not rely on the viewport height (-95%)
* not rely on fixing the highlighted item after click
* support both scrolling up and scrolling down
* support fast scrolling

SleeplessByte commented:
![RTL implementation of the highlight feature](https://github.com/rails/rails/assets/1964376/fa58a254-8887-4161-b3c6-1ac3d6a03db2)

The nav highlight is now also supported in RTL mode, shouldn't be flaky with flick/fast scrolling, and should now work both going up and down. 

@AmandaPerino ready for final review.

p8 commented:
Thanks for sticking with this @SleeplessByte !
It seems the side bar no longer scrolls when the active element is outside the visible region:
https://github.com/rails/rails/assets/28561/ca82b356-e8b7-4223-93a8-3c42b4b1ce40

p8 commented:
Is the red highlight maybe a bit too much in light mode?

Current version:
<img width="873" alt="image" src="https://github.com/rails/rails/assets/28561/e4f375df-97fe-419f-86c0-b06bcbd03e0f">

This PR:
<img width="873" alt="image" src="https://github.com/rails/rails/assets/28561/b306d9be-c3e3-4fab-a734-f8365548f018">

SleeplessByte commented:
> Thanks for sticking with this @SleeplessByte !
> It seems the side bar no longer scrolls when the nav highlight is outside the visible region:
> https://github.com/rails/rails/assets/28561/ca82b356-e8b7-4223-93a8-3c42b4b1ce40
> 
> I'm also getting the following error (maybe related):
> > Uncaught ReferenceError: backToTop is not defined
>     toggleBackToTop .../rails/guides/output/javascripts/guides.js:113
> 
> 

Yes, the scrolling of the sidebar stops scrolling on windows, so we need to implement that differently but i want to get this over the line before we do. See the comment in the code.

The second thing is a merge artifact. Will fix!

SleeplessByte commented:
> Is the white text with red highlight maybe a bit too much in light mode?
> 
> Current version:
> <img width="862" alt="image" src="https://github.com/rails/rails/assets/28561/05c2294a-9153-444a-bcd7-fc521b79b55d">
> 
> This PR:
> <img width="870" alt="image" src="https://github.com/rails/rails/assets/28561/9dca9ca7-5692-42a0-9019-cbc6d01e41dc">
> 
> 
> 
> 

Because the links are red, I dont think it's smart to invert it, but an off-white may help to reduce the contrast a bit.

SleeplessByte committed:
Apply review commentary

p8 commented:
> Yes, the scrolling of the sidebar (`scrollTo`, `scrollTop` etc), stops scrolling on Windows, which means that as you are scrolling the main element, and the side element is scrolled, it completely stops scrolling the main element. So we need to implement that differently but I want to get this PR over the line before we do. See the comment in the code to understand more.

This happens to me on OSX, both in FireFox and Chrome. I'm not allowed to merge code (only documentation) but I'd be a bit hesitant to merge this while the scrolling of the sidebar is a bit broken.

I know you put a lot of time and energy into this and this is all volunteer work.
Not sure how comfortable you are with git and I don't want put more work on your plate.
Would it be possible to split the sidebar changes from the other changes?
Or could we easily cherry-pick some of the earlier commits?

> The second thing is a merge artifact. Will fix!

Thanks. This has been fixed.

SleeplessByte commented:
This is broken on the live site / main if you have smooth scrolling on, so this PR does not break it.

I'm okay with trying to find a solution to the issue instead if that is what's keeping this from being merged. Why don't I do that before cherry picking the commits (which I also know how to do).

SleeplessByte commented:
(to respond to my comment about windows: I meant the implementation before this PR.)

p8 commented:
> (to respond to my comment about windows: I meant the implementation before this PR.)

Ah, that makes sense. Thanks!

SleeplessByte commented:
Still on the list, I think I have a solution. Will try to test it some more this week.

p8 commented:
@SleeplessByte It would be nice if we could merge some of these changes.
We are getting [PR's ](https://github.com/rails/rails/pull/52838)to fix issues addressed by this PR.

SleeplessByte commented:
Yap.

Ping me again on Friday if I haven't pushed yet.

p8 commented:
Thanks @SleeplessByte !

SleeplessByte commented:
Was able to empty my friday out so I'll work on this tomorrow!

SleeplessByte committed:
Merge branch 'main' into accessibility

SleeplessByte committed:
Fix issues with scrolling behaviour

To "restore" the current behaviour on main, a solution was needed
for browser-OS combinations that do not allow for two concurrent
scrolling elements. Current code on main breaks completely when
using scroll-behaviour: smooth (in CSS), and breaks when scrolling
back-to-top from the bottom of the page (highlight no longer
correct).

This commit works around the limitation whilst also enabling the
non-smooth scrolled solution for those who want reduced motion.
Additionally it has a workaround for a bug on Android, and some
defaults for those without JavaScript enabled.

SleeplessByte commented:
@p8 your concern has been addressed.

SleeplessByte committed:
Ensure chapters list is always scrolled after scroll-end

p8 commented:
Thanks @SleeplessByte, this is looking great!
@carlosantoniodasilva do you think this could be merged?

p8 commented:
@zzak Do you think this is good to go?

zzak commented:
@SleeplessByte Could you squash your commits please? I will try to make time to review this soon. :bow:

SleeplessByte commented:
@zzak squashing will lose a lot of valuable information for anyone later looking at the individual lines of code plus git blame. Almost each commit has a detailed message of the whys.

If possible I'd rather not, so to make future maintenance easier.

zzak commented:
Ok, I can see you went to great lengths to provide additional context in each commit, and since this is not touching Rails framework code it maybe acceptable as-is but our policy is to always squash commits before merging.

I'm still catching up on stuff, have been sick so apologies for the delay. Another good person to ask might be @seanpdoyle if they can give this a look :bow:

SleeplessByte commented:
> our policy is to always squash commits before merging.

Yep, hence me asking for the exception in the original PR message 🗡️ 
Hoping this can be looked at soon though, I think these changes are impactful for many users.

SleeplessByte committed:
Merge branch 'main' into accessibility

zzak commented:
Ok I had some time to at least check the changes listed above, especially the darkmode and screen reader stuff, definitely an improvement on the UX.

Two things stood out to me.

![Screenshot from 2024-10-18 20-13-41](https://github.com/user-attachments/assets/9a4bbdd2-20ab-4ac4-b52d-9bef136a0783)

The "skip to article body" link covers the nav and it feels like a bug.

![Screenshot from 2024-10-18 20-21-35](https://github.com/user-attachments/assets/823e163a-1b14-4b1c-95c0-5918608ded54)

The guides selector is blank when viewing on a page in a small browser window.

I didn't take a hard look at the code, just checking if there were any obvious traps. I'm not a big fan of the cosmetic changes, like whitespace, semi-colons, etc -- just because it makes the diff much larger than it needs to be. These types of changes should be delegated to a linter which we can easily auto-format for us and get everyone on the same page -- but we haven't needed one for the guides until now so I'm not sure. There are also multiple "merge commits" on the branch that I'd like to see rebased out, but these are just some very minor knits.

seanpdoyle reviewed (approved):
I agree with @zzak that the volume of changes related to indentation, the addition of `;` characters, and changes from `_` characters to `-` characters made reviewing this diff more challenging than it would be otherwise. Having said that, viewing the diff [ignoring spacing changes](https://github.com/rails/rails/pull/51499/files?w=1) improved that experience.

Changes to the document's structure, elements' semantics, and form controls' accessible names are all improvements. I'm less sure about the particulars of the JavaScript implementation, but that feedback might be better reserved for a separate pull request discussion.

SleeplessByte commented:
> The "skip to article body" link covers the nav and it feels like a bug.

It's common practice. I am personally very used to skiplinks. However, given that this can be entirely controlled by CSS, perhaps an improvement could be addressed in a future PR?

> The guides selector is blank when viewing on a page in a small browser window.

I saw that too. If okay, I'd like to address that in a follow-up PR myself.

> I'm not a big fan of the cosmetic changes, [...]

Wholeheartedly agree with everything you said here @zzak. In fact, in my response to an earlier review on this PR I said the same. This could all be prevented if there was an auto-formatter (not linter). The wide variation of styles in the same JavaScript file threw me off a lot when trying to dissect how to change it. Personally I do think you needed it, given the disjoint style it is in on main. My two cents? It should be added! And then we can all agree that we don't agree, but the system decides, and auto-formatting wins all!

> There are also multiple "merge commits" on the branch that I'd like to see rebased out

I've rebased this PR 6 times now, and they are not easy because of a lot of conflicts with `main`. 
The "merge" commits you see (https://github.com/rails/rails/pull/51499/commits/43b8eb88df42309647c28b8ca59570163aa7a410 and https://github.com/rails/rails/pull/51499/commits/3d389b01c0ec534b465de836ab723fa16b9b4324) are likely conflict resolution commits. TBH they're only here because this has been kept on the shelve since I first did the work for so long. If this is an absolute must, I can do it, but if the PR then sits for another few months accumulating conflicts, I'll likely no longer be interested, as I don't know if I can find the will to go through the interactive process again. If one of the maintainers happens to be a star with pulling commits out (without the full rebase), this PR *is* open for edits! 

> Having said that, viewing the diff [ignoring spacing changes](https://github.com/rails/rails/pull/51499/files?w=1) improved that experience.

Great! I tend to use the ignore whitespace setting myself as well. I don't understand the `;` comment though. I've followed the style of the file which was semis yes, so the only places where it now has an `;` it was missing it before. As for `-` vs `_` in the names/ids, I tried to be consistent. It's discussed earlier in the PR. My screenreader separates words better when hyphenated than when underscored, but I do realise this feels like noise to the PR.

----------------------------------------------------

I am fine with a follow-up issue I'm assigned to to address issues or improvements (in smaller bits, so easier to review perhaps?) if that's something that would help get this merged.

Earlopain commented:
I believe this PR has caused issues on smaller screens with the chapter select: https://github.com/rails/rails/issues/53500. When the chapter select changes to take up the entire screen width, it is impossible to scroll down.

brunoprietog commented:
Actually, the reason behind using the JavaScript code there was because we only wanted to set the roles when we have JavaScript enabled. If that's not a concern and we'll assume that will always be the case, it's probably fine.
